### PR TITLE
feat(varlen): support varlen training for huggingface models

### PR DESCRIPTION
**Pull request by zigzagcai**
## Motivation

support varlen training for huggingface models

## Modification

support varlen training for huggingface models

Diff for modelling_internlm.py and modelling_internlm2.py:
https://github.com/InternLM/InternEvo-HFModels/commit/e1f604e54691af412ece13ef103ef7ebc5c68598

## BC-breaking (Optional)

None

## Use cases (Optional)

None

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Review by sallyjunjun — APPROVED** [review]

**Comment by sallyjunjun** [comment]
后续需要将dispatch路径中，模型相关的代码剥离到InternEvo-HFModels代码仓

**Comment by zigzagcai** [comment]
> 后续需要将dispatch路径中，模型相关的代码剥离到InternEvo-HFModels代码仓

~~这个PR还涉及其它地方配套的改动，所以需要等这个PR merge后，再单独把dispatch放到代码仓。~~

Update:
dispatch代码已经删掉了，改为手动修改huggingface modelling文件来支持变长训练

File: internlm/train/pipeline.py [hunk 1]
```diff
@@ -144,21 +144,10 @@ def _check_module(name, module):
...
-            elif gpc.is_initialized(ParallelMode.TENSOR) and not is_using_isp():
-                setattr(param, IS_TENSOR_ZERO_PARALLEL, True)
-
     for _chunk in unwrap_naive_amp(model):
         # set param parallel attribute
         for name, module in _chunk.named_modules():
-            if gpc.config.model_type == "hf":
-                _check_module_hf(name, module)
-            else:
-                _check_module(name, module)
```
Comment (sallyjunjun) [inline internlm/train/pipeline.py]: 这里使用_check_module能覆盖所有参数的属性吗？因为在这个函数里，会严格判断module是RMSNorm，Embedding1D等，而用户自定义的modeling文件里面模块会不一样。

Comment (zigzagcai) [reply internlm/train/pipeline.py]: 在dispatch逻辑里，我们自动替换了，基础的ops都复用InternEvo的ops，即nn.Embedding - > Embedding1D, nn.Linear -> new_linear(), norm -> new_norm()，就可以完全覆盖这些基础ops的parallel attributes情况

**Comment by zigzagcai** [comment]
Done.

We choose to apply manual code changes in huggingface models to support varlen training.

**Comment by zigzagcai** [comment]
Diff for modelling_internlm.py and modelling_internlm2.py:
https://github.com/InternLM/InternEvo-HFModels/commit/e1f604e54691af412ece13ef103ef7ebc5c68598

File: internlm/train/pipeline.py (hunk 1 above)
Comment (zigzagcai) [reply internlm/train/pipeline.py]: Update:
dispatch代码已经删掉了，参数属性手动根据huggingface modelling的情况，case by case来提供

**Comment by zigzagcai** [comment]
Done.

**Review by sunpengsdu — APPROVED** [review]

**Pull request merged**